hydrogym/firedrake/utils/io.py

```diff
@@ -82,6 +83,38 @@ def __call__(self, iter: int, t: float, flow: Tuple[fd.Function]):
         print(self.print_fmt.format(*new_data.ravel()))
 
 
+def log_postprocess(flow):
+  obs = flow.evaluate_objective(return_objective_values=True)
+  if isinstance(obs, collections.abc.Iterable):
+    return obs
+  else:
+    return [obs]
+
+
+class RewardCallback(CallbackBase):
+
+  def __init__(
+      self,
+      num_sim_substeps_per_actuation: int,
+      postprocess: Callable = log_postprocess,
+      interval: Optional[int] = 1,
+  ):
+    super().__init__(interval=interval)
+    self.postprocess = postprocess
+    self.num_sim_substeps_per_actuation = num_sim_substeps_per_actuation
+    self.data = []
+
+  def __call__(self, iter: int, t: float, flow: Tuple[fd.Function]):
+    if iter % self.interval == 0:
+      # self.data.append([self.postprocess(flow)])
+      self.data.append([*self.postprocess(flow)])
+      # if ((iter + 1) * self.dt * 10e6) % (self.DT * 10e6) == 0:
+      if (iter + 1) % self.num_sim_substeps_per_actuation == 0:
+        self.data = np.array(self.data)
+        flow.reward_array = self.data
```

Comment:
Is there another way to do this?  The flow is a physics model and so RL-specific stuff should be abstracted